srtcore/api.cpp

```diff
@@ -1908,11 +1908,29 @@ int srt::CUDTUnited::close(const SRTSOCKET u)
         return 0;
     }
 #endif
-    CUDTSocket* s = locateSocket(u);
-    if (!s)
-        throw CUDTException(MJ_NOTSUP, MN_SIDINVAL, 0);
+#if ENABLE_HEAVY_LOGGING
+    // Wrapping the log into a destructor so that it
+    // is printed AFTER the destructor of SocketKeeper.
+    struct ForceDestructor
+    {
+        CUDTSocket* ps;
+        ForceDestructor(): ps(NULL){}
+        ~ForceDestructor()
+        {
+            if (ps) // Could be not acquired by SocketKeeper, occasionally
+            {
+                HLOGC(smlog.Debug, log << "CUDTUnited::close/end: @" << ps->m_SocketID << " busy=" << ps->isStillBusy());
+            }
+        }
+    } fod;
+#endif
+
+    SocketKeeper k(*this, u, ERH_THROW);
+    IF_HEAVY_LOGGING(fod.ps = k.socket);
```

Comment:
Please find a better name, e.g. `LogOnDestruct` or `ScopedLog`.

```suggestion
    struct ScopedLog
    {
        const CUDTSocket* const ps;
        ScopedLog(const CUDTSocket* psock): ps(psock){}
        ~ScopedLog()
        {
            if (ps) // Could be not acquired by SocketKeeper, occasionally
            {
                HLOGC(smlog.Debug, log << "CUDTUnited::close/end: @" << ps->m_SocketID << " busy=" << ps->isStillBusy());
            }
        }
    };
#endif

    SocketKeeper k(*this, u, ERH_THROW);
    IF_HEAVY_LOGGING(ScopedLog slog(k.socket));
```

Reply:
I named it `ScopedExitLog`, as its intention is to print a log line just before exiting.